BUILD.bazel

```diff
@@ -0,0 +1,379 @@
+load("//:bazel/flex.bzl", "genlex")
+load("//:bazel/bison.bzl", "genyacc")
+
+package(
+    default_visibility = ["//visibility:public"],
+    licenses = ["notice"],  # Apache v2
+)
+
+filegroup(
+    name = "p4include",
+    srcs = glob(["p4include/*.p4"]),
+)
+
+P4C_BUILD_DEFAULT_COPTS = [
+    # Where p4c should look for p4include at runtime.
+    ("-DCONFIG_PKGDATADIR=\\\"external/%s\\\"" % repository_name()),
+    # This will work only if the binary is executed by Bazel. For a general
+    # solution, we would need to make p4c aware of Bazel, specifically:
+    # https://github.com/bazelbuild/bazel/blob/master/tools/cpp/runfiles/runfiles_src.h
+]
+
+genrule(
+    name = "sed_config_h",
+    srcs = ["cmake/config.h.cmake"],
+    outs = ["config.h"],
+    cmd = ("sed 's|cmakedefine|define|g' $< | " +
+           "sed 's|define HAVE_LIBGC 1|undef HAVE_LIBGC|g' > $@"),
+    visibility = ["//visibility:private"],
+)
+
+cc_library(
+    name = "config_h",
+    hdrs = [
+        "config.h",
+    ],
+    include_prefix = ".",
+    visibility = [":__subpackages__"],
+)
+
+genyacc(
+    name = "p4_parser_yacc",
+    src = "frontends/parsers/p4/p4parser.ypp",
+    extra_outs = ["frontends/parsers/p4/stack.hh"],
+    header_out = "frontends/parsers/p4/p4parser.hpp",
+    source_out = "frontends/parsers/p4/p4parser.cc",
+)
+
+genyacc(
+    name = "v1_parser_yacc",
+    src = "frontends/parsers/v1/v1parser.ypp",
+    extra_outs = ["frontends/parsers/v1/stack.hh"],
+    header_out = "frontends/parsers/v1/v1parser.hpp",
+    source_out = "frontends/parsers/v1/v1parser.cc",
+)
+
+genrule(
+    name = "p4lexer_lex",
+    srcs = ["frontends/parsers/p4/p4lexer.ll"],
+    outs = ["frontends/parsers/p4/p4lexer.lex"],
+    cmd = ("sed '/%option outfile=\"lex.yy.c\"/d' $< > $@"),
+    visibility = ["//visibility:private"],
+)
+
+genlex(
+    name = "p4lexer",
+    src = "frontends/parsers/p4/p4lexer.lex",
+    out = "frontends/parsers/p4/p4lexer.cc",
+    prefix = "yy",
+    visibility = ["//visibility:private"],
+)
+
+genrule(
+    name = "v1lexer_lex",
+    srcs = ["frontends/parsers/v1/v1lexer.ll"],
+    outs = ["frontends/parsers/v1/v1lexer.lex"],
+    cmd = ("sed '/%option outfile=\"lex.yy.c\"/d' $< > $@"),
+    visibility = ["//visibility:private"],
+)
+
+genlex(
+    name = "v1lexer",
+    src = "frontends/parsers/v1/v1lexer.lex",
+    out = "frontends/parsers/v1/v1lexer.cc",
+    prefix = "yy",
+    visibility = ["//visibility:private"],
+)
+
+cc_library(
+    name = "p4c_includes",
+    includes = [
+        "ir/",
+        "lib/",
+        "tools/ir-generator",
+    ],
+    visibility = [":__subpackages__"],
+)
+
+# This rule helps split some circular dependencies between subdirectories.
```

Comment:
If we have circular dependencies this may be a bug, perhaps we should investigate.
Could you file an issue describing the dependency you found?

Reply:
This comment originates from the Stratum authors, so I don't know the details here.